test/integration/supervisor_tls_spec_test.go

```diff
@@ -0,0 +1,266 @@
+			expectedError:      `The OIDCIdentityProvider "invalid-oidc-idp-invalid-kind" is invalid: spec.tls.certificateAuthorityDataSource.kind: Unsupported value: "": supported values: "Secret", "ConfigMap"`,
+		},
+		{
+			name: "should disallow certificate authority data source with invalid kind",
+			customResourceYaml: []byte(here.Doc(`
+			---
+			apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+			kind: OIDCIdentityProvider
+			metadata:
+				name: invalid-oidc-idp-invalid-kind
+			spec:
+				tls:
+					certificateAuthorityDataSource:
+						kind: sorcery
+						name: foo
+						key: bar
+				issuer: https://foo.bar.com/oauth2/default
+				authorizationConfig:
+					additionalScopes: [offline_access, email]
+					allowPasswordGrant: true
+				client:
+					secretName: foo-bar-client-credentials
+			`)),
+			customResourceName: "invalid-oidc-idp-invalid-kind",
+			expectedError:      `The OIDCIdentityProvider "invalid-oidc-idp-invalid-kind" is invalid: spec.tls.certificateAuthorityDataSource.kind: Unsupported value: "sorcery": supported values: "Secret", "ConfigMap"`,
+		},
+		{
+			name: "should create a custom resource passing all validations using a Secret source",
+			customResourceYaml: []byte(here.Doc(`
+			---
+			apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+			kind: OIDCIdentityProvider
+			metadata:
+				name: valid-oidc-idp-secret-kind1
+			spec:
+				tls:
+					certificateAuthorityDataSource:
+						kind: Secret
+						name: foo
+						key: bar
+				issuer: https://foo.bar.com/oauth2/default
+				authorizationConfig:
+					additionalScopes: [offline_access, email]
+					allowPasswordGrant: true
+				client:
+					secretName: foo-bar-client-credentials
+			`)),
+			customResourceName: "valid-oidc-idp-secret-kind1",
+			expectedError:      "",
+		},
+		{
+			name: "should create a custom resource passing all validations using a ConfigMap source",
+			customResourceYaml: []byte(here.Doc(`
+			---
+			apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+			kind: OIDCIdentityProvider
+			metadata:
+				name: valid-oidc-idp-cm-kind1
+			spec:
+				tls:
+					certificateAuthorityDataSource:
+						kind: ConfigMap
+						name: foo
+						key: bar
+				issuer: https://foo.bar.com/oauth2/default
+				authorizationConfig:
+					additionalScopes: [offline_access, email]
+					allowPasswordGrant: true
+				client:
+					secretName: foo-bar-client-credentials
+			`)),
+			customResourceName: "valid-oidc-idp-cm-kind1",
+			expectedError:      "",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			resourceName := tc.customResourceName
+			expectedError := tc.expectedError
+			yamlBytes := tc.customResourceYaml
+			t.Parallel()
+			yamlFilepath := filepath.Join(t.TempDir(), fmt.Sprintf("tls-spec-validation-%s.yaml", resourceName))
+
+			require.NoError(t, os.WriteFile(yamlFilepath, yamlBytes, 0600))
+			cmd := exec.CommandContext(context.Background(), "kubectl", []string{"apply", "-f", yamlFilepath}...)
```

Comment:
Using `testutil.testlib.CreateTestOIDCIdentityProvider` (and other similar test helpers for the other resource types) would have some advantages over kubectl. The spec is provided as Go structs, which is nice for Go tests. They are automatically put into the right namespace, they are automatically named, they are automatically cleaned up, they automatically check their phase, etc. Unless maybe there is a specific reason to use kubectl.

Reply:
Similarly, using the Golang client code to create and delete resources is easier for a Go test, unless there is a specific reason to use kubeectl for the test.

Reply:
These test ensures that the validation in the API spec is working as expected and any future changes don’t accidentally violate the current set of assumptions/preconditions. Reconciliation of these resources is not what is being tested. But the test utility function always expects the resource creation to be successful.. https://github.com/vmware-tanzu/pinniped/blob/main/test/testlib/client.go#L657-L661